### PR TITLE
Refactor sparse unary ops tests

### PR DESCRIPTION
**Pull request by vfdev-5**
Description:
- Added TestSparseUnaryUfuncs test case to test_sparse.py
- Removed old test code for neg, asin, log1p ops on sparse data
- Added sparse unary ops testing helpers in OpInfo, e.g. sample_sparse_inputs
- Added abs, sign UnaryUfuncInfo to `op_db`

cc @mruberry

### CODE REVIEW

File: test/test_sparse.py [hunk 1]
```diff
@@ -3080,5 +2861,60 @@ def test_cuda_sparse_cpu_dense_add(self):
...
+            self.assertEqual(expected_output, op.operator_variant(sparse_tensor_copy).to_dense())
+
+    def _test_sparse_unary_op(self, sparse_inputs, device, dtype, op):
+        for sparse_input in sparse_inputs:
+            sparse_tensor = sparse_input.input[0]
+            self._test_sparse_unary_op_on_input(sparse_tensor, device, dtype, op)
+
+            for a_op in op.aliases:
+                # Add required attributes from base op
+                a_op.supports_tensor_out = op.supports_tensor_out
```
Comment (vfdev-5) [inline test/test_sparse.py:2902]: This is a drawback of introducing AliasInfo vs reusing UnaryUfuncInfo

Comment (mruberry) [reply test/test_sparse.py:2902]: I think this will be OK, though, since aliases should be tested rarely.

**Comment by vfdev-5** [comment]
Related broken tests:
- [test_reference_numerics_abs_cpu_int8 (__main__.TestUnaryUfuncsCPU) ... /var/lib/jenkins/workspace/aten/src/ATen/native/Math.h:1108:10: runtime error: 128 is outside the range of representable values of type 'signed char'](https://app.circleci.com/pipelines/github/pytorch/pytorch/249112/workflows/85d4c618-8577-4ac0-aba5-512ba61e5a70/jobs/9516819)

Unrelated broken tests (?):
- https://app.circleci.com/pipelines/github/pytorch/pytorch/249112/workflows/85d4c618-8577-4ac0-aba5-512ba61e5a70/jobs/9516818
- https://app.circleci.com/pipelines/github/pytorch/pytorch/249112/workflows/85d4c618-8577-4ac0-aba5-512ba61e5a70/jobs/9516921
- https://app.circleci.com/pipelines/github/pytorch/pytorch/249112/workflows/85d4c618-8577-4ac0-aba5-512ba61e5a70/jobs/9517147
- https://app.circleci.com/pipelines/github/pytorch/pytorch/249112/workflows/85d4c618-8577-4ac0-aba5-512ba61e5a70/jobs/9516955
- https://app.circleci.com/pipelines/github/pytorch/pytorch/249112/workflows/85d4c618-8577-4ac0-aba5-512ba61e5a70/jobs/9516758

**Comment by mruberry** [comment]
Failures are all in the base, sorry about that @vfdev-5. We can ignore them and I've reverted the PR causing them.

File: test/test_sparse.py [hunk 2]
```diff
@@ -3237,15 +3014,21 @@ def test_cuda_sparse_cpu_dense_add(self):
...
 
     @ops(sparse_unary_ufuncs)
-    def test_sparse_consistency(self, device, dtype, op):
+    def test_sparse_consistency_dense_to_sparse(self, device, dtype, op):
+
         unsupportedTypes = [torch.bfloat16, torch.cfloat, torch.cdouble]
         if dtype in unsupportedTypes:
             self.skipTest('Skipped! Unsupported dtypes for Sparse')
 
+        if device == "cpu" and dtype == torch.float16:
```
Comment (mruberry) [inline test/test_sparse.py:3028]: Instead of `if device == "cpu"` use `if self.device_type == 'cpu'` or `if torch.device(device).type == 'cpu'`. We prefer the latter approaches since `if device == "cuda"` won't work, since there are multiple CUDA devices and their names are rendered like "cuda:0" and "cuda:1".

File: test/test_sparse.py [hunk 3]
```diff
@@ -3263,7 +3046,85 @@ def test_sparse_consistency(self, device, dtype, op):
...
+                self._test_sparse_unary_op_on_input(sparse_tensor, device, dtype, a_op)
+
+    @ops(sparse_unary_ufuncs)
+    def test_sparse_consistency_sparse_to_dense(self, device, dtype, op):
+
+        if dtype in [torch.bfloat16, torch.bool, torch.complex64, torch.complex128]:
+            # to_dense() will call coalesce() -> RuntimeError: "coalesce" not implemented
+            self.skipTest('Skipped! "coalesce" not implemented for {}'.format(dtype))
+
+        if dtype in [torch.half, ] and device == "cpu":
```
Comment (mruberry) [inline test/test_sparse.py:3096]: Same device check here as mentioned above. Also prefer using tuples instead of lists for iterables in Python. Only use a list if the iterable has to be mutable

File: torch/testing/_internal/common_methods_invocations.py [hunk 4]
```diff
@@ -80,6 +81,20 @@ def _getattr_qual(obj, name, default=_NOTHING):
...
+    class AliasInfo(object):
+        """Placeholder for aliases. For example, torch.abs -> 
+        torch.absolute, torch.Tensor.absolute, torch.Tensor.absolute_
+        """
+
+        def __init__(self, alias_name):
+            self.op = _getattr_qual(torch, alias_name)
+            self.method_variant = getattr(torch.Tensor, alias_name, None)
+            self.inplace_variant = getattr(torch.Tensor, alias_name + "_", None)
+            self.operator_variant = getattr(operator, alias_name, None)
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:93]: I think we can do without the operator variant

File: torch/testing/_internal/common_methods_invocations.py [hunk 5]
```diff
@@ -80,6 +81,20 @@ def _getattr_qual(obj, name, default=_NOTHING):
 class OpInfo(object):
     """Operator information and helper functions for acquiring it."""
 
+    class AliasInfo(object):
+        """Placeholder for aliases. For example, torch.abs ->
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:85]: Instead of "Placeholder for aliases", which is ambiguous, I think this can just say that it acquires and holds alias information.

File: torch/testing/_internal/common_methods_invocations.py [hunk 6]
```diff
@@ -108,7 +123,9 @@ def __init__(self,
                  promotes_integers_to_float=False,  # whether op promotes unary output to float or not
                  sample_inputs_func=None,  # function to generate sample inputs
                  aten_name=None,  # name of the corresponding aten:: operator
-                 supports_sparse=False  # supported for sparse
+                 aliases=None,  # list of aliases, e.g. torch.absolute -> torch.abs
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:126]: "list" -> "iterable"

We want to prefer tuples instead of lists, since tuples are immutable, but any iterable will do

File: torch/testing/_internal/common_methods_invocations.py [hunk 7]
```diff
@@ -108,7 +123,9 @@ def __init__(self,
...
+                 supports_sparse=False,  # whether the op correctly handles sparse input
+                 sparse_op_info=None,  # sparse op info as dict, e.g. flag that op supports inplace on uncoalesced
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:128]: Instead of "sparse_op_info" I think we can just inline the sparse properties like any OpInfo property. So a property like `supports_inplace_on_uncoalesced` would just be any other property.

If the number of sparse-specific properties gets too large then we can revisit having a separate SparseOpInfo class with a pointer to the original OpInfo inside it.

File: torch/testing/_internal/common_methods_invocations.py [hunk 8]
```diff
@@ -108,7 +123,9 @@ def __init__(self,
                  promotes_integers_to_float=False,  # whether op promotes unary output to float or not
                  sample_inputs_func=None,  # function to generate sample inputs
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:124]: I would expect "sparse_sparse_inputs_func" to appear here (note slight name change for consistency)

File: torch/testing/_internal/common_methods_invocations.py [hunk 9]
```diff
@@ -148,8 +165,11 @@ def __init__(self,
         else:
             self.autodiff_nonfusible_nodes = autodiff_nonfusible_nodes
         self.supports_sparse = supports_sparse
-
-
+        self.sparse_op_info = sparse_op_info if sparse_op_info is not None else {}
+        self.operator_variant = getattr(operator, name, None)
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:169]: Adding operator variants is interesting but we should probably separate it into its own PR

File: torch/testing/_internal/common_methods_invocations.py (hunk 6 above)
Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:126]: The comment should give an example of the correct way to add an alias, which would be `("absolute",)` for `torch.abs`, right?

File: torch/testing/_internal/common_methods_invocations.py [hunk 10]
```diff
@@ -148,8 +165,11 @@ def __init__(self,
...
+        self.aliases = []
+        if aliases is not None:
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:171]: Adding alias variants like this is also very interesting but we should probably separate it into its own PR, too

File: torch/testing/_internal/common_methods_invocations.py [hunk 11]
```diff
@@ -171,10 +191,20 @@ def get_inplace(self):
...
+        """Returns operator variant of the operator, e.g. operator.neg
+        Returns None if the operator has no operator variant.
+        """
+        return self.operator_variant
+
     def sample_inputs(self, device, dtype, requires_grad=False):
         """Returns an iterable of SampleInputs."""
         return self.sample_inputs_func(self, device, dtype, requires_grad)
 
+    def sample_sparse_inputs(self, device, dtype, requires_grad=False, **kwargs):
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:204]: See name change note above.

I think this can call `self.sparse_sample_inputs_func` analogous to `sample_inputs` and just return `None` if there is no such function.

File: torch/testing/_internal/common_methods_invocations.py [hunk 12]
```diff
@@ -171,10 +191,20 @@ def get_inplace(self):
...
+        """Returns an iterable of SampleInputs."""
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:205]: This docstring should clarify the intent of these SampleInputs, as opposed to sample_inputs() above, is for use in testing the sparse version of these operations.

File: torch/testing/_internal/common_methods_invocations.py [hunk 13]
```diff
@@ -280,6 +311,60 @@ def __init__(self,
         #   outside a function's domain.
         self._domain_eps = 1e-5
 
+    def sample_sparse_inputs(self, device, dtype, requires_grad=False, **kwargs):
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:314]: I think we can reuse the same pattern that sample_inputs() uses for the UnaryUfuncs? See comment above.

File: torch/testing/_internal/common_methods_invocations.py [hunk 14]
```diff
@@ -280,6 +311,60 @@ def __init__(self,
...
+        """Returns an iterable of SampleInputs."""
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:315]: Let's elaborate in this docstring on what kind of sparse tensors are generated.

File: torch/testing/_internal/common_methods_invocations.py [hunk 15]
```diff
@@ -280,6 +311,60 @@ def __init__(self,
...
+        if not self.supports_sparse:
+            raise RuntimeError("This op does not support sparse input")
+
+        low, high = self.domain
+        low = -10 if low is None else low + self._domain_eps
+        high = 10 if high is None else high - self._domain_eps
+        is_coalesced = kwargs.get("is_coalesced", False)
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:322]: I would expect sparse sample inputs for the unary ufuncs to generate a variety of tensors, including coalesce, uncoalesced and hybrid tensors. Is it really interesting to continue separating them? I always though that separation in the current sparse test suite was an artifact of how the tests are generated.

Put another way: the sample inputs are supposed to be the sample inputs, we shouldn't intend to partition them based on different kwargs.

File: torch/testing/_internal/common_methods_invocations.py [hunk 16]
```diff
@@ -280,6 +311,60 @@ def __init__(self,
...
+
+        numel = 10
+        t1 = torch.sparse_coo_tensor(
+            indices=torch.arange(numel).unsqueeze(0),
+            values=torch.linspace(low, high, steps=numel),
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:327]: We should probably generate random values like the unary ufuncs do for strided

File: torch/testing/_internal/common_methods_invocations.py [hunk 17]
```diff
@@ -280,6 +311,60 @@ def __init__(self,
...
+        # empty sparse tensor
+        t3 = torch.sparse_coo_tensor(
+            indices=torch.zeros([2, 0]),
+            values=torch.zeros([0, 5, 5, 5, 5, 5, 5, 0]),
+            size=[0, 0, 5, 5, 5, 5, 5, 5, 0],
+            device=device,
+            dtype=dtype
+        )
+
+        if is_coalesced:
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:356]: See above comment. Let's always generate all the relevant sample inputs (and document which are generated in the docstring).

File: torch/testing/_internal/common_methods_invocations.py [hunk 18]
```diff
@@ -280,6 +311,60 @@ def __init__(self,
...
+            raise RuntimeError("This op does not support sparse input")
+
+        low, high = self.domain
+        low = -10 if low is None else low + self._domain_eps
+        high = 10 if high is None else high - self._domain_eps
+        is_coalesced = kwargs.get("is_coalesced", False)
+
+        numel = 10
+        t1 = torch.sparse_coo_tensor(
+            indices=torch.arange(numel).unsqueeze(0),
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:326]: 1D is interesting, but 2D with a density target of ~20% is probably more "representative"

File: torch/testing/_internal/common_methods_invocations.py [hunk 19]
```diff
@@ -108,7 +123,9 @@ def __init__(self,
                  promotes_integers_to_float=False,  # whether op promotes unary output to float or not
                  sample_inputs_func=None,  # function to generate sample inputs
                  aten_name=None,  # name of the corresponding aten:: operator
-                 supports_sparse=False  # supported for sparse
+                 aliases=None,  # list of aliases, e.g. torch.absolute -> torch.abs
+                 supports_sparse=False,  # whether the op correctly handles sparse input
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:127]: If suppost_sparse is inferable from the presence of a sparse_sample_inputs function then we should infer the property in the constructor instead of requiring the user pass it correctly.

I think it's inferable?

File: torch/testing/_internal/common_methods_invocations.py [hunk 20]
```diff
@@ -280,6 +311,60 @@ def __init__(self,
...
+
+        numel = 10
+        t1 = torch.sparse_coo_tensor(
+            indices=torch.arange(numel).unsqueeze(0),
+            values=torch.linspace(low, high, steps=numel),
+            size=[numel, ],
+            device=device,
+            dtype=dtype,
+        )
+        # hybrid sparse input
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:332]: We may want to leave the addition of hybrid sparse tensors to the sample_inputs as a TODO. I'd prefer we focus on generating the non-hybrid tensors properly. In particular, it seems like it'd be nice to have a function like `make_tensor`, `make_sparse_tensor` which could produce them. The creation of sparse tensors with different properties today seems very ad-hoc and challenging.

Producing `make_sparse_tensor` could be a follow-up PR to this one, but then I would especially emphasize keeping this function simple for now until `make_sparse_tensor` is produced.

File: torch/testing/_internal/common_methods_invocations.py [hunk 21]
```diff
@@ -1020,6 +1111,50 @@ def sample_inputs_pinverse(op_info, device, dtype, requires_grad=False):
                                 dtypes=[torch.bfloat16])),
                    promotes_integers_to_float=True,
                    handles_complex_extremals=False),
+    UnaryUfuncInfo('abs',
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:1114]: Adding these OpInfos is great but we should separate their addition. 

Maybe the OpInfo for abs can go in the PR adding alias support to OpInfos? (That PR can also update test_op_aliases.py.)

File: torch/testing/_internal/common_methods_invocations.py [hunk 22]
```diff
@@ -1020,6 +1111,50 @@ def sample_inputs_pinverse(op_info, device, dtype, requires_grad=False):
...
+                       SkipInfo('TestCommon', 'test_variant_consistency_eager',
+                                dtypes=[torch.cfloat, torch.cdouble]),
+                       SkipInfo('TestCommon', 'test_variant_consistency_jit',
+                                dtypes=[torch.cfloat, torch.cdouble, torch.bfloat16]),
+                   ),
+                   test_inplace_grad=False,
+                   assert_autodiffed=True,
+                   aliases=['absolute', ],
+                   supports_sparse=False),  # Temporary disabled until https://github.com/pytorch/pytorch/pull/43330
+    UnaryUfuncInfo('sign',
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:1141]: This should be separated into its own PR, especially since it doesn't even support sparse. Note that OpInfos should be inserted alphabetically into this iterable.

File: test/test_sparse.py [hunk 23]
```diff
@@ -3263,7 +3046,85 @@ def test_sparse_consistency(self, device, dtype, op):
...
+
+        if len(sparse_inputs) == 0:
+            self.skipTest("Skipped! No sample inputs!")
+
+        self._test_sparse_unary_op(sparse_inputs, device, dtype, op)
+
+    @ops(sparse_unary_ufuncs)
+    def test_sparse_consistency_sparse_to_dense_coalesced(self, device, dtype, op):
+
+        if dtype in [torch.bfloat16, torch.bool, torch.complex64, torch.complex128]:
```
Comment (mruberry) [inline test/test_sparse.py:3111]: This PR may want to refactor these checks into a helper function. We do something like this in test_ops.py, for example:

https://github.com/pytorch/pytorch/blob/46cf6d332f075ed90d3baf21c32de51e4f304549/test/test_ops.py#L106

File: test/test_sparse.py [hunk 24]
```diff
@@ -3263,7 +3046,85 @@ def test_sparse_consistency(self, device, dtype, op):
...
+        # op is common_methods_invocations.UnaryUfuncInfo
+        sparse_inputs = op.sample_sparse_inputs(device, dtype)
+
+        if len(sparse_inputs) == 0:
+            self.skipTest("Skipped! No sample inputs!")
+
+        self._test_sparse_unary_op(sparse_inputs, device, dtype, op)
+
+    @ops(sparse_unary_ufuncs)
+    def test_sparse_consistency_sparse_to_dense_coalesced(self, device, dtype, op):
```
Comment (mruberry) [inline test/test_sparse.py:3109]: We should probably give up on the granularity of coalesced or not and unify these two entrypoints and one helper into a single test.

File: test/test_sparse.py [hunk 25]
```diff
@@ -3263,7 +3046,85 @@ def test_sparse_consistency(self, device, dtype, op):
...
+        if op.operator_variant is not None:
+            sparse_tensor_copy = sparse_tensor.clone()
+            self.assertEqual(expected_output, op.operator_variant(sparse_tensor_copy).to_dense())
+
+    def _test_sparse_unary_op(self, sparse_inputs, device, dtype, op):
+        for sparse_input in sparse_inputs:
+            sparse_tensor = sparse_input.input[0]
+            self._test_sparse_unary_op_on_input(sparse_tensor, device, dtype, op)
+
+            for a_op in op.aliases:
```
Comment (mruberry) [inline test/test_sparse.py:3083]: This alias-specific testing can be moved into its own PR.

Follow-up question: should we test that aliases work properly for sparse? I can see arguments both ways. If we do, however, we should probably have a special test for alias consistency where only the first sample input is given to the alias and the forward and backward are compared with the original op.

File: test/test_sparse.py [hunk 26]
```diff
@@ -3263,7 +3046,85 @@ def test_sparse_consistency(self, device, dtype, op):
         assert torch.is_tensor(output)
         self.assertEqual(output.to_dense(), expected)
 
+    def _test_sparse_unary_op_on_input(self, sparse_tensor, device, dtype, op):
```
Comment (mruberry) [inline test/test_sparse.py:3049]: This is where we can apply more granularity.

We can have separate tests for function, method, inplace, and out=.

The function variant of the test should go through all sample inputs and verify that forward/backward is equivalent to dense (the dense gradient will need to be masked out to be compared, and a helper function might be useful for this).

The method variant should just run on a single sample input and compare forward/backward to the function variant. The method variant is very unlikely to have diveregent behavior from the function variant.

The inplace variant is trickier. It needs special handling (like you already have below) and its gradient often needs even more special handling. I could see passing all sample inputs to it and performing a series of thorough checks. We may even need to add a new property like "supports_sparse_inplace_grad" to OpInfo.

File: test/test_sparse.py [hunk 27]
```diff
@@ -3237,15 +3014,21 @@ def test_cuda_sparse_cpu_dense_add(self):
         with self.assertRaisesRegex(RuntimeError, "add: expected 'self' to be a CUDA tensor, but got a CPU tensor"):
             x + sparse_y
 
+
 class TestSparseUnaryUfuncs(TestCase):
     exact_dtype = True
 
     @ops(sparse_unary_ufuncs)
-    def test_sparse_consistency(self, device, dtype, op):
+    def test_sparse_consistency_dense_to_sparse(self, device, dtype, op):
```
Comment (mruberry) [inline test/test_sparse.py:3022]: This test should probably be refactored into the new, more granular functions described below.

**Review by mruberry — COMMENTED** [review]
Hey @vfdev-5! 

This is really cool. I think we can restructure it to simplify the design and make it more consistent with the strided tests, however. Also, this PR has a lot of good ideas, including improved variant testing and alias testing. We should probably separate those into other PRs. What I expect from this PR is:

- OpInfos add support for sparse sample inputs, and these sample inputs are generated in an interesting way
- the sparse test suite is updated to compare forward and backward of these inputs on the functional variant of the sparse op vs. the function variant of the strided op (this may require another helper or two)

Then we can look at improving on this base with adding support for aliases, adding support for operator variants, and improve sparse test coverage. I would really like to see alias support for OpInfos, in particular, so we can kill test_op_aliases.py, which is a pain to extend.

File: torch/testing/_internal/common_methods_invocations.py (hunk 15 above)
Comment (vfdev-5) [reply torch/testing/_internal/common_methods_invocations.py:322]: I agree. Let's put everything inside a single list. The reason for partition the input is that during the test we can see on which data the test is failing...

Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:322]: I understand there's a tension between test granularity and this approach. I think we should solve the granularity issue differently however, by adding support for named subtests, for example, or naming the sample inputs and throwing more informative failure messages.

File: test/test_sparse.py [hunk 28]
```diff
@@ -3248,33 +3025,146 @@ def test_cuda_sparse_cpu_dense_add(self):
...
-    def test_sparse_consistency(self, device, dtype, op):
-        unsupportedTypes = [torch.bfloat16, torch.cfloat, torch.cdouble]
+    def _skip_helper(self, dtype):
+        unsupportedTypes = [torch.bfloat16, torch.float16, torch.bool, torch.complex64, torch.complex128]
         if dtype in unsupportedTypes:
-            self.skipTest('Skipped! Unsupported dtypes for Sparse')
+            self.skipTest(f'Skipped! "coalesce" not implemented for {dtype}')
+
+    def _get_dense_and_sparse_samples(self, device, dtype, op):
+        test_backward = op.test_complex_grad or not dtype.is_complex
```
Comment (mruberry) [inline test/test_sparse.py:3038]: We also have supports_autograd now:

https://github.com/pytorch/pytorch/blob/e3a805b9c5fceeb1bd58d7cd309271f513386595/torch/testing/_internal/common_methods_invocations.py#L164

File: test/test_sparse.py [hunk 29]
```diff
@@ -3248,33 +3025,146 @@ def test_cuda_sparse_cpu_dense_add(self):
...
+        samples = op.sample_inputs(device, dtype, requires_grad=test_backward)
+        sparse_samples = op.sparse_sample_inputs(device, dtype, requires_grad=test_backward)
```
Comment (mruberry) [inline test/test_sparse.py:3040]: So in this model sparse_sample_inputs() are ADDITIONAL sparse-specific inputs?

File: test/test_sparse.py [hunk 30]
```diff
@@ -3248,33 +3025,146 @@ def test_cuda_sparse_cpu_dense_add(self):
...
+        if sparse_samples is not None:
+            samples += sparse_samples
+        return samples
+
+    def _test_forward(self, sample, op):
+        dense_sample_input = tuple(
```
Comment (mruberry) [inline test/test_sparse.py:3046]: This assumes that args and kwargs don't need to be modified between dense and sparse inputs, right? We should make these assumptions explicit (in a comment).

File: test/test_sparse.py [hunk 31]
```diff
@@ -3248,33 +3025,146 @@ def test_cuda_sparse_cpu_dense_add(self):
...
+        samples = self._get_dense_and_sparse_samples(device, dtype, op)
+
+        if len(samples) == 0:
+            self.skipTest("Skipped! No sample inputs!")
+
+        inplace = op.get_inplace()
+        if inplace is None:
+            self.skipTest("Skipped! No inplace variant op")
+
+        for sample in samples:
```
Comment (mruberry) [inline test/test_sparse.py:3135]: Add a comment here about why this is necessary. 

Couldn't the sample input be copied (or deepcopied) and then passed to _test_forward() to reuse more of the helper code?

File: test/test_sparse.py [hunk 32]
```diff
@@ -3248,33 +3025,146 @@ def test_cuda_sparse_cpu_dense_add(self):
...
+                (i.to_dense() if i.layout != torch.strided else i) for i in sample.input
+            )
+            sparse_sample_input = tuple(
+                (i.to_sparse() if i.layout == torch.strided else i) for i in sample.input
+            )
+            assert len(sparse_sample_input) == 1, "Unary op should have single input"
+
+            expected_forward = op(*dense_sample_input, *sample.args, **sample.kwargs)
+
+            if op.supports_inplace_on_uncoalesced or sparse_sample_input[0].is_coalesced():
```
Comment (mruberry) [inline test/test_sparse.py:3146]: Comment explaining this behavior, too. Inplace operations are tricky.

File: test/test_sparse.py [hunk 33]
```diff
@@ -3248,33 +3025,146 @@ def test_cuda_sparse_cpu_dense_add(self):
...
+                    sparse_forward = inplace(*(clone_input_helper(input) for input in sparse_sample_input),
+                                             *sample.args,
+                                             **sample.kwargs)
+                    self.assertEqual(sparse_forward.to_dense(), expected_forward)
+                else:
+                    with self.assertRaisesRegex(RuntimeError, "result type cannot be"):
+                        sparse_forward = inplace(*(clone_input_helper(input) for input in sparse_sample_input),
+                                                 *sample.args,
+                                                 **sample.kwargs)
+            else:
```
Comment (mruberry) [inline test/test_sparse.py:3157]: Comment for what this else-clause covers

File: torch/testing/_internal/common_methods_invocations.py [hunk 34]
```diff
@@ -158,11 +158,12 @@ def __init__(self,
                  decorators=None,  # decorators to apply to generated tests
                  safe_casts_outputs=False,  # whether op allows safe casting when writing to out arguments
                  sample_inputs_func=None,  # function to generate sample inputs
+                 sparse_sample_inputs_func=None,  # function to generate sparse sample inputs
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:161]: See above comment. These sparse sample inputs are used IN ADDITION to the regular sample inputs?

Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:161]: Also we should clarify that setting this is what indicates the op supports sparse inputs

File: torch/testing/_internal/common_methods_invocations.py [hunk 35]
```diff
@@ -293,6 +304,31 @@ def sample_inputs_unary(op_info, device, dtype, requires_grad):
...
+    Generated values are bound to domain if op_info.domain is provided.
+    Generated COO matrices are coalesced and uncoalesced of random shape.
+    """
+    if not op_info.supports_sparse:
+        raise RuntimeError("This op does not support sparse input")
+
+    # generate 2D random coalesced/uncoalesced sparse matrix with density of 20%
+    rh, rw = torch.randint(100, 200, size=(2, )).tolist()
+    low, high = op_info.domain
+    domain = None if (low is None and high is None) else (low + op_info._domain_eps, high - op_info._domain_eps)
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:319]: This seems cool. In the future should we consider updating the sparse matrix ctor to be make_sparse_tensor() for consistency with make_tensor(), and look at giving them similar interfaces and logic?

File: torch/testing/_internal/common_methods_invocations.py [hunk 36]
```diff
@@ -1463,8 +1500,7 @@ def sample_inputs_masked_select(op_info, device, dtype, requires_grad):
                    aliases=('arcsin', ),
                    ref=np.arcsin,
                    domain=(-1, 1),
-                   supports_sparse=True,
-                   decorators=(precisionOverride({torch.bfloat16: 1e-2}),),
+                   decorators=(precisionOverride({torch.bfloat16: 1e-2, torch.float16: 5e-4}),),
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:1503]: What is the half precision change needed?

File: torch/testing/_internal/common_utils.py [hunk 37]
```diff
@@ -1728,6 +1728,10 @@ def random_sparse_matrix(rows, columns, density=0.01, **kwargs):
     dtype = kwargs.get('dtype', torch.double)
     device = kwargs.get('device', 'cpu')
     singular = kwargs.get("singular", False)
+    values_domain = kwargs.get("domain", None)
```
Comment (mruberry) [inline torch/testing/_internal/common_utils.py:1731]: In the future we should also add tests to test_testing to ensure that random_sparse_matrix() (and make_tensor()) work as expected

File: torch/testing/_internal/common_utils.py [hunk 38]
```diff
@@ -1736,12 +1740,40 @@ def random_sparse_matrix(rows, columns, density=0.01, **kwargs):
     column_indices = [i % columns for i in range(nonzero_elements)]
     random.shuffle(column_indices)
     indices = [row_indices, column_indices]
-    values = torch.randn(nonzero_elements, dtype=dtype, device=device)
-    # ensure that the diagonal dominates
-    values *= torch.tensor([-float(i - j)**2 for i, j in zip(*indices)], dtype=dtype, device=device).exp()
+    if dtype in integral_types():
```
Comment (mruberry) [inline torch/testing/_internal/common_utils.py:1743]: This domain will be incorrect for uint8

File: torch/testing/_internal/common_utils.py [hunk 39]
```diff
@@ -1736,12 +1740,40 @@ def random_sparse_matrix(rows, columns, density=0.01, **kwargs):
...
+        requires_grad = False
```
Comment (mruberry) [inline torch/testing/_internal/common_utils.py:1744]: For the construction of values, should this defer to make_tensor()?

File: torch/testing/_internal/common_utils.py (hunk 37 above)
Comment (mruberry) [inline torch/testing/_internal/common_utils.py:1731]: Why is this function **kwargs? 

We don't need to fix random_sparse_matric() in this PR but that seems odd, and it makes documenting its supported kwargs tricky.

File: torch/testing/_internal/common_utils.py [hunk 40]
```diff
@@ -1736,12 +1740,40 @@ def random_sparse_matrix(rows, columns, density=0.01, **kwargs):
...
+            high = min(high, 100)  # if high=inf, clamp to 100
+            values = (high - low) * torch.rand(nonzero_elements, dtype=dtype, device=device) + low
+
     indices_tensor = torch.tensor(indices)
-    A = torch.sparse_coo_tensor(indices_tensor, values, (rows, columns), device=device)
-    return A.coalesce()
+    A = torch.sparse_coo_tensor(
+        indices_tensor, values, (rows, columns), device=device, requires_grad=requires_grad
+    )
+    if not uncoalesced:
```
Comment (mruberry) [inline torch/testing/_internal/common_utils.py:1774]: In the future I'd like to better consider how to generate sparse tensors, especially how to handle coalescing (or not) and the density target

File: torch/testing/_internal/common_methods_invocations.py [hunk 41]
```diff
@@ -293,6 +304,31 @@ def sample_inputs_unary(op_info, device, dtype, requires_grad):
...
+def sparse_sample_inputs_unary(op_info, device, dtype, requires_grad):
+    """This function generates 2D sparse COO matrices with 20% density. 
+    Generated values are bound to domain if op_info.domain is provided.
+    Generated COO matrices are coalesced and uncoalesced of random shape.
+    """
+    if not op_info.supports_sparse:
+        raise RuntimeError("This op does not support sparse input")
+
+    # generate 2D random coalesced/uncoalesced sparse matrix with density of 20%
+    rh, rw = torch.randint(100, 200, size=(2, )).tolist()
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:317]: These don't need to be randomly generated, just pick values

**Review by mruberry — COMMENTED** [review]
Hey @vfdev-5!

Overall this looks pretty good. I made some inline comments/questions, mostly requesting more comments in the code to explain the choices/assumptions we're making. In the future we probably want to look more closely at generating sparse tensors, too.

File: torch/testing/_internal/common_methods_invocations.py (hunk 36 above)
Comment (vfdev-5) [reply torch/testing/_internal/common_methods_invocations.py:1503]: It is a failure on cuda:
```
test_function_consistency_asin_cuda_float16

AssertionError: False is not true : Tensors failed to compare as equal!With rtol=0.001 and atol=1e-05, found 1 element(s) (out of 29925) whose difference(s) exceeded the margin of error (including 0 nan comparisons). The greatest difference was 0.00048828125 (-0.216064453125 vs. -0.216552734375), which occurred at index (53, 37).  ```

Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:1503]: It's fine as is, I was just curious.

File: test/test_sparse.py (hunk 30 above)
Comment (vfdev-5) [reply test/test_sparse.py:3046]: Yes, it is rather strong assumption I was making as that the test case is for unary ops only (class name TestSparseUnaryUfuncs) and for which args/kwargs wont be tensors.
Anyway, I'll put that into a comment.

**Review by mruberry — APPROVED** [review]
Nice step forward for sparse testing, @vfdev-5!
